src/commands/registerCommands.ts

```diff
@@ -179,24 +180,81 @@ async function grammarAssociationCommand(documentURI: Uri, languageClient: Langu
   const fileUri = await window.showOpenDialog(options);
   if (fileUri && fileUri[0]) {
     // The XSD, DTD has been selected, get the proper syntax for binding this grammar file in the XML document.
-    const identifier = TextDocumentIdentifier.create(documentURI.toString());
     const grammarURI = fileUri[0];
     try {
-      const result = await commands.executeCommand(ServerCommandConstants.ASSOCIATE_GRAMMAR_INSERT, identifier, grammarURI.toString(), bindingType);
-      // Insert the proper syntax for binding
-      const lspTextDocumentEdit = <TextDocumentEdit>result;
-      const workEdits = new WorkspaceEdit();
-      for (const edit of lspTextDocumentEdit.edits) {
-        workEdits.replace(documentURI, languageClient.protocol2CodeConverter.asRange(edit.range), edit.newText);
+      if (bindingType == 'fileAssociation') {
+        // Bind grammar using file association
+        await bindWithFileAssociation(documentURI, grammarURI);
+      } else {
+        // Bind grammar using standard binding
+        await bindWithStandard(documentURI, grammarURI, bindingType, languageClient);
       }
-      workspace.applyEdit(workEdits); // apply the edits
-
     } catch (error) {
       window.showErrorMessage('Error during grammar binding: ' + error.message);
     };
   }
 }
 
+/**
+ * Perform grammar binding using file association through settings.json
+ *
+ * @param documentURI the URI of the current XML document
+ * @param grammarURI the URI of the user selected grammar file
+ */
+async function bindWithFileAssociation(documentURI: Uri, grammarURI: Uri) {
+  const identifier = TextDocumentIdentifier.create(documentURI.toString());
+  const grammarFilename = path.basename(grammarURI.toString())
```

Comment:
It doesn't work when you select a DTD/XSD which is not in the same folder. The main idea is to store:

 * 1) the URI file path:
 * 
```json
{
      "pattern": "**/bar.xml",
      "systemId": "file:///c%3A/Users/azerr/Downloads/vscode-xml-catalog-test/catalog-preferences/foo.dtd"
    }
```

 * 2) and if the selected XSD/DTD comes from the workspace we should store the absolute path from workspace root.
```json
{
      "pattern": "**/bar.xml",
      "systemId": "/catalog-preferences/foo.dtd"
    }
```

To support 1), replace with 

`const grammarFilePath = grammarURI.toString();`